### PR TITLE
[WIP][RFC] Fastly support

### PR DESCRIPTION
**Pull request by hpatoio**
Adding support for Fastly 
Goes with https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/pull/427

For details about Fastly API:

 - https://docs.fastly.com/api/purge#soft_purge_2e4d29085640127739f8467f27a5b549
 - https://docs.fastly.com/api/purge#purge_d8b8e8be84c350dd92492453a3df3230 

Related discussion: #142

### CODE REVIEW

**Comment by dbu** [comment]
thanks for starting this! i am super busy before php benelux, will try to have a look at this next week.

do you have specific questions at this point? what remains to be done (apart from test and style issues)?

**Comment by hpatoio** [comment]
> do you have specific questions at this point? 

No, not really

> what remains to be done (apart from test and style issues)?

The FastlyProxyClient need implements:

- [x] TagCapable
- [ ] PurgeCapable
- [ ] RefreshCapable

And then documentation.

**Comment by dbu** [comment]
cool. then i will ignore that pull request for now, please ping me when you need input or want me to review something. i would love to have fastly support in this library!

File: src/ProxyClient/Fastly.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            'Fastly-Key' => $this->options['authentication_token'],
+            'Accept' => 'application/json'
+        ];
+
+        if (true === $this->options['soft_purge']) {
+            $headers['Fastly-Soft-Purge'] = 1;
+        }
+
+        foreach (\explode(",", $this->options['service_identifier']) as $fastlyServiceId) {
+            foreach ($tags as $tag) {
```
Comment (andrerom) [inline src/ProxyClient/Fastly.php]: In the case of hard purge, it's possible to invalidate 256 tags at a time now:
https://docs.fastly.com/api/purge#purge_db35b293f8a724717fcf25628d713583

**Comment by andrerom** [comment]
I think we might have another todo here. From what I could see last week, UserContextInvalidator also needs to be changed to not rely on BanCable for this to be possible.

One alternative would be to add tags for user id and session id on the hash request cache. Any thoughts on this @dbu ?

**Comment by dbu** [comment]
so it would be possible to do user context caching with fastly, if we find a way around the ban issue? that would be quite cool indeed.

how do we do that with the symfony cache kernel thingy? that one does not have full banning implemented, but does have tags. do we not invalidate the hash lookup for that case?

**Comment by dbu** [comment]
tagging the hash lookup response with the session id would be a very good idea imho. with the x-cache extension, invalidating the tag would also be much better for varnish than the ban request that we currently do. (apart from that ban request being completely broken and banning the whole server each time)

**Comment by Toflar** [comment]
The Symfony Proxy is not `BanCapable` but I'm not sure what the question here is?

**Comment by andrerom** [comment]
> The Symfony Proxy is not BanCapable but I'm not sure what the question here is?

Q: How Use Context hash invalidation is done with Symfony Proxy, given [UserContextInvalidator](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/src/UserContextInvalidator.php#L39) in bundle expects BanCapable, which Symfony Proxy does not implement.

At least I got issue with that when trying to do [Fastly POC](https://github.com/andrerom/sf-london-2018-httpcache-demo/commit/747a2eb4b004b2133526b4cafb92648226dd6d81#diff-3584dcd2a732e6f34941b9194c2ff416R109) using `cache_manager.custom_proxy_client`.

---

@dbu While on that, I also ran into another possible thing we should think about for full Fastly support. And that is mapping FOS's `reverse_proxy_ttl ` header to Fastly's `Surrogate-Control` header: https://docs.fastly.com/guides/performance-tuning/controlling-caching#changing-caching-times-for-different-users

**Comment by Toflar** [comment]
I'm not using the User Context at all, I'm using our https://github.com/terminal42/header-replay-bundle. So I think this probably doesn't work?

**Comment by andrerom** [comment]
+100

**Comment by andrerom** [comment]
> do you plan to work on any of the subjects we discussed here @andrerom to wrap up fastly support?

I havn't planned for that, [what I did for demo at symfony live can be found here](https://github.com/andrerom/sf-london-2018-httpcache-demo/commit/63d3f0a2148d7f10f3bfd9da951ff4b61f9aa8f4). It's not tested against Fastly, but it has support for purging several tags at once when doing hard purging which can also be added here, but besides that it's very similar to the code here.

File: src/ProxyClient/Fastly.php (hunk 1 above)
Comment (andrerom) [reply src/ProxyClient/Fastly.php]: Seems they are also able to do this on Soft purge, via support there they seem to plan to update doc to make it more clear.

So given Faslty has API rate limits, we should change to use the API linked to above regardless of soft/purge so we can do up to 256 at a time.

**Comment by dbu** [comment]
continued in #451

**Pull request closed without merging**